scipy/stats/_sensitivity_analysis.py

```diff
@@ -0,0 +1,634 @@
+    qualitative analysis. Nevertheless, when the number of input parameters
+    increases such analysis becomes unrealistic as it would be difficult to
+    conclude on high-order terms. Hence the benefit of using Sobol' indices.
+
+    """
+    random_state = check_random_state(random_state)
+
+    n_ = int(n)
+    if not (n_ & (n_ - 1) == 0) or n != n_:
+        raise ValueError(
+            "The balance properties of Sobol' points require 'n' "
+            "to be a power of 2."
+        )
+    n = n_
+
+    if not callable(method):
+        indices_methods: Dict[str, Callable] = {
+            "saltelli_2010": saltelli_2010,
+        }
+        try:
+            method = method.lower()  # type: ignore[assignment]
+            indices_method = indices_methods[method]
+        except KeyError as exc:
+            message = (
+                f"{method!r} is not a valid 'method'. It must be one of"
+                f" {set(indices_methods)!r} or a callable."
+            )
+            raise ValueError(message) from exc
+    else:
+        kind = inspect.Parameter.POSITIONAL_OR_KEYWORD
+        parameters = (
+            inspect.Parameter('f_A', kind=kind),
+            inspect.Parameter('f_B', kind=kind),
+            inspect.Parameter('f_AB', kind=kind)
+        )
+        sig = inspect.Signature(parameters=parameters)
+        indices_method = method
+
+        if inspect.signature(indices_method) != sig:
+            message = (
+                "If 'method' is a callable, it must have the following"
+                f" signature: {inspect.signature(saltelli_2010)}"
+            )
+            raise ValueError(message)
+
+    if callable(func):
+        if dists is None:
+            raise ValueError(
+                "'dists' must be defined when 'func' is a callable."
+            )
+
+        A, B = sample_A_B(n=n, dists=dists, random_state=random_state)
+        AB = sample_AB(A=A, B=B)
+
+        f_A = np.atleast_2d(func(A))
+
+        if f_A.shape[1] != n or f_A.shape[0] < 1:
+            raise ValueError(
+                "'func' output should have a shape ``(s, -1)`` with ``s`` "
+                "the number of output."
+            )
+
+        f_B, f_AB = np.atleast_2d(func(B), func(AB))
+    else:
+        message = (
+            "When 'func' is a dictionary, it must contain the following "
+            "keys: 'f_A', 'f_B' and 'f_AB'."
+            "'f_A' and 'f_B' should have a shape ``(s, n)`` and 'f_AB' "
+            "should have a shape ``(s, d*n)``."
+        )
+        try:
+            f_A, f_B, f_AB = np.atleast_2d(
+                func['f_A'], func['f_B'], func['f_AB']
+            )
+        except KeyError as exc:
+            raise ValueError(message) from exc
+
+        if f_A.shape[1] != n or f_A.shape != f_B.shape or \
+                f_AB.shape == f_A.shape or f_AB.shape[1] % n != 0:
+            raise ValueError(message)
+
+    # Compute indices
```

Comment:
This is Saltelli's estimator, but it is numerically unstable. The correct way to implement the estimator is to remove the sample mean of the estimator. This is because the variance is independent of a constant which is added to the function value. This was known to Sobol' :

![image](https://user-images.githubusercontent.com/31351465/211209427-6042dff5-04ac-447c-b515-1c416e4022ad.png)

This fact is mentioned in (Sobol’ & Levitan, 1998) and (Sobol’ & Levitan, 1999):

![image](https://user-images.githubusercontent.com/31351465/211209468-d3fca4e3-9d40-4d14-bde5-efdb0edba7e1.png)

This was only much later that we discovered how true this is (Baudin et al. 2016). This is similar to the issue of estimating the sample variance (see (Chan, Golub, & LeVeque, 1983)). This is how the formula is implemented in OpenTURNS:

![image](https://user-images.githubusercontent.com/31351465/211209383-fe3f8cad-095a-4456-bb90-9823453a3876.png)

This is not vanilla Saltelli's estimator, but it is the right way to estimate Sobol' indices. 

References:
-----------------
- Sobol, I. M. (1993). Sensitivity analysis for non-linear mathematical models. Mathematical modelling and computational experiment, 1 :407–414.
- Sobol’, I. and Levitan, Y. L. (1998). On the use of variance reducing multipliers in monte carlo computations of a
global sensitivity index. In K.Chan, Tarantola, S., and Campolongo, F., editors, Second Symposium on
theory and applications of sensitivity analysis of model output in computer
simulation (SAMO).
- Sobol’, I. and Levitan, Y. L. (1999). On the use of variance reducing multipliers in monte carlo computations of a
global sensitivity index. Computer Physics Communications, 117(1) :52–61.
- Baudin, M., Boumhaout, K., Delage, T., Iooss, B., and Martinez, J. (2016). Numerical stability of sobol’ indices estimation formula. 8th International Conference on SAMO, Saint Pierre, La Réunion, France.
- Chan, Tony F., Gene H. Golub, and Randall J. LeVeque. "Algorithms for computing the sample variance: Analysis and recommendations." The American Statistician 37.3 (1983): 242-247.